File: Assignment 1/question4/question4.c
#include <stdio.h>

int is_valid_solution(int *values) {
    // Έλεγχος ότι κάθε ψηφίο είναι διαφορετικό
    for (int i = 0; i < 10; i++) {
        for (int j = i + 1; j < 10; j++) {
            if (values[i] == values[j]) {
                return 0; // Μη έγκυρη λύση
            }
        }
    }

    // Έλεγχος εάν η πρόσθεση είναι έγκυρη
    int base = values['A' - 'A'];
    int ball = values['B' - 'A'];
    int games = values['G' - 'A'];
    int total1 = base * 1000 + ball * 100 + ball * 10 + base;
    int total2 = games * 10000 + ball * 1000 + ball * 100 + base * 10 + games;
    
    return total1 + total1 == total2;
}

void solve_cryptarithm(int *values, int index) {
    if (index == 10) { // Έχουμε βρει μια πιθανή λύση
        if (is_valid_solution(values)) {
            printf("A = %d, B = %d, E = %d, G = %d, L = %d, M = %d, S = %d\n",
                   values['A' - 'A'], values['B' - 'A'], values['E' - 'A'],
                   values['G' - 'A'], values['L' - 'A'], values['M' - 'A'],
                   values['S' - 'A']);
        }
        return;
    }

    for (int i = 0; i < 10; i++) {
        values[index] = i;
        solve_cryptarithm(values, index + 1);
    }
}

int main() {
    int values[10];
    solve_cryptarithm(values, 0);
    return 0;
}

File: Assignment 1/question4/question4.py
def solve_cryptarithmetic(puzzle, mapping, used_digits):
    if len(mapping) == len(puzzle):  # Όταν η αντιστοίχιση είναι πλήρης
        # Έλεγχος αν η αντιστοίχιση οδηγεί σε έγκυρη λύση
        if sum(mapping[c] for c in puzzle[:-1]) == mapping[puzzle[-1]]:
            return mapping  # Επιστρέφει την αντιστοίχιση
        else:
            return None  # Αν δεν είναι έγκυρη, επιστρέφει None
    else:
        next_char = next(filter(lambda x: x not in mapping, puzzle))  # Επιλογή του επόμενου χαρακτήρα που πρέπει να αντιστοιχηθεί
        for digit in range(10):  # Δοκιμή κάθε ψηφίου από 0 έως 9
            if digit not in used_digits:  # Έλεγχος αν το ψηφίο έχει ήδη χρησιμοποιηθεί
                mapping[next_char] = digit  # Αντιστοίχιση του επόμενου χαρακτήρα με το ψηφίο
                used_digits.add(digit)  # Προσθήκη του ψηφίου στα ψηφία που έχουν χρησιμοποιηθεί
                solution = solve_cryptarithmetic(puzzle, mapping, used_digits)  # Αναδρομική κλήση για τον επόμενο χαρακτήρα
                if solution is not None:  # Αν βρέθηκε λύση
                    return solution
                else:  # Αν δεν βρέθηκε λύση, επαναφορά των αλλαγών
                    del mapping[next_char]
                    used_digits.remove(digit)
        return None  # Αν δεν βρέθηκε λύση με κανένα ψηφίο, επιστρέφει None


def solve_puzzle(puzzle):
    mapping = {}  # Αρχικοποίηση αντιστοίχισης
    used_digits = set()  # Αρχικοποίηση συνόλου χρησιμοποιημένων ψηφίων
    chars = set(puzzle)  # Οι μοναδικοί χαρακτήρες στο παζλ
    chars.remove('+')  # Αφαίρεση του χαρακτήρα '+'
    chars.remove('=')  # Αφαίρεση του χαρακτήρα '='
    chars = list(chars)  # Μετατροπή του συνόλου σε λίστα για να διατηρήσουμε τη σειρά
    return solve_cryptarithmetic(puzzle, mapping, used_digits)


# Κύριο πρόγραμμα για δοκιμή
if __name__ == "__main__":
    puzzle = "BASE + BALL = GAMES"
    solution = solve_puzzle(puzzle)
    if solution:
        print("Solution found:")
        for char, digit in solution.items():
            print(f"{char}: {digit}")
    else:
        print("No solution found")
